### ISSUE TITLE
IAM auth doesn't work with a service account on a Compute Engine VM

### ISSUE THREAD

**Issue by samsaam-thg** [googleapis/langchain-google-alloydb-pg-python#110]
Trying to run my job in GCP batch with a service account attached to the job. I have confirmed the SA is attached, I added a debug print statement to the `_get_iam_principal_email` function and confirmed the credentials has the service account email attribute (why does it not just return at this point instead of calling the oauth endpoint?) https://github.com/googleapis/langchain-google-alloydb-pg-python/blob/1c6b90af0fc3c97f747c43649b88d90f75a32bce/src/langchain_google_alloydb_pg/alloydb_engine.py#L63
```
    if hasattr(credentials, "_service_account_email"):
        email = credentials._service_account_email
        print(email)
```

To reproduce:
```
from langchain_google_alloydb_pg import AlloyDBEngine

engine = AlloyDBEngine.from_instance(
        project_id=config.project_id,
        region=config.region,
        cluster=config.alloydb_cluster,
        instance=config.cluster_instance,
        database=config.instance_database,
        ip_type=config.alloydb_ip_type,
)
```

Relevant library versions:
langchain-google-alloydb-pg==0.1.0
google==3.0.0
google-api-core==2.18.0
google-auth==2.29.0
google-cloud-alloydb-connector==0.4.1
google-cloud-core==2.4.1
google-cloud-storage==2.16.0
google-crc32c==1.5.0
google-resumable-media==2.7.0
googleapis-common-protos==1.63.0
aiohttp==3.9.5

**Comment by samsaam-thg** [comment]
### Full Reproducibility Instructions

When running the following auth_debug.py locally, impersonating the service account by:
`export GOOGLE_APPLICATION_CREDENTIALS=/path/to/keyfile.json`

This seems to work.

However, when running in a docker container on a batch job (on a VM), it fails to find the email. Presumably this is because $GOOGLE_APPLICATION_CREDENTIALS env variable is not set, and it uses the metadata server to acquire the credentials.

To recreate:
requirements.txt
```
aiohttp==3.9.5
aiosignal==1.3.1
annotated-types==0.6.0
anyio==4.3.0
asn1crypto==1.5.1
async-timeout==4.0.3
asyncpg==0.29.0
attrs==23.2.0
backports.tarfile==1.1.1
beautifulsoup4==4.12.3
cachetools==5.3.3
certifi==2024.2.2
cffi==1.16.0
charset-normalizer==3.3.2
cryptography==42.0.5
dataclasses-json==0.6.4
Deprecated==1.2.14
distro==1.9.0
filelock==3.13.4
frozenlist==1.4.1
fsspec==2024.3.1
google==3.0.0
google-api-core==2.18.0
google-auth==2.28.0
google-cloud-alloydb-connector==1.0.0
google-cloud-core==2.4.1
google-cloud-storage==2.16.0
google-crc32c==1.5.0
google-resumable-media==2.7.0
googleapis-common-protos==1.63.0
greenlet==3.0.3
grpcio==1.62.2
grpcio-tools==1.62.2
h11==0.14.0
html2text==2024.2.26
httpcore==1.0.5
httpx==0.27.0
huggingface-hub==0.22.2
idna==3.7
importlib_metadata==7.1.0
jaraco.classes==3.4.0
jaraco.context==5.3.0
jaraco.functools==4.0.1
jeepney==0.8.0
Jinja2==3.1.3
joblib==1.4.0
jsonpatch==1.33
jsonpointer==2.4
keyring==25.1.0
keyrings.google-artifactregistry-auth==1.1.2
langchain==0.1.9
langchain-community==0.0.24
langchain-core==0.1.27
langchain-google-alloydb-pg==0.2.0
langchain-openai==0.0.6
langsmith==0.1.49
MarkupSafe==2.1.5
marshmallow==3.21.1
more-itertools==10.2.0
mpmath==1.3.0
multidict==6.0.5
mypy-extensions==1.0.0
networkx==3.3
numpy==1.26.4
openai==1.23.2
orjson==3.10.1
packaging==23.2
pg8000==1.31.1
pgvector==0.2.5
pillow==10.3.0
pluggy==1.5.0
proto-plus==1.23.0
protobuf==4.25.3
pyasn1==0.6.0
pyasn1_modules==0.4.0
pycparser==2.22
pydantic==2.7.1
pydantic_core==2.18.2
PyJWT==2.8.0
python-dateutil==2.9.0.post0
PyYAML==6.0.1
regex==2024.4.16
requests==2.31.0
rsa==4.9
safetensors==0.4.3
scikit-learn==1.4.2
scipy==1.13.0
scramp==1.4.5
SecretStorage==3.3.3
sentence-transformers==2.4.0
six==1.16.0
sniffio==1.3.1
soupsieve==2.5
SQLAlchemy==2.0.29
sympy==1.12
tenacity==8.2.3
threadpoolctl==3.4.0
tiktoken==0.6.0
tokenizers==0.19.1
tqdm==4.66.2
transformers==4.40.0
typing-inspect==0.9.0
typing_extensions==4.11.0
urllib3==2.2.1
wrapt==1.16.0
yarl==1.9.4
zipp==3.18.1
```

auth_debug.py
```
import google.auth
import google.auth.transport.requests
from typing import Dict
import aiohttp
import asyncio


async def _get_iam_principal_email(credentials) -> str:
    if not credentials.valid:
        print("Credentials not valid")
        request = google.auth.transport.requests.Request()
        credentials.refresh(request)
    if hasattr(credentials, "_service_account_email"):
        print("Has attr triggered")
        email = credentials._service_account_email
        print(email)
    # call OAuth2 api to get IAM principal email associated with OAuth2 token
    url = f"https://oauth2.googleapis.com/tokeninfo?access_token={credentials.token}"
    async with aiohttp.ClientSession() as client:
        response = await client.get(url, raise_for_status=True)
        response_json: Dict = await response.json()
        print(response_json)
        email = response_json.get("email")
    if email is None:
        raise ValueError(
            "Failed to automatically obtain authenticated IAM principal's "
            "email address using environment's ADC credentials!"
        )
    return email.replace(".gserviceaccount.com", "")

credentials, _ = google.auth.default(scopes=["https://www.googleapis.com/auth/userinfo.email"])
email = asyncio.run(_get_iam_principal_email(credentials))
print(email)
```
Build an image with the following Dockerfile:

```Dockerfile
FROM python:3.11.8-slim

WORKDIR /app/

COPY requirements.txt ./
COPY auth_debug.py ./

RUN pip install -r requirements.txt

ENTRYPOINT ["python", "auth_debug.py"]

```

Build the image however you please and upload to artifact registry.

Submit a job:

batch.json
```json
{
    "taskGroups": [
        {
            "taskSpec": {
                "runnables": [
                    {
                        "container": {
                            "imageUri": "europe-docker.pkg.dev/my-project/docker/my-image:latest"
                        }
                    }
                ],
                },
                "computeResource": {
                    "cpuMilli": 2000,
                    "memoryMib": 2000,
                    "bootDiskMib": 10000
                },
                "maxRetryCount": 0,
                "maxRunDuration": "3600s"
            },
            "taskCount": 1,
            "parallelism": 1
        }
    ],
    "logsPolicy": {
        "destination": "CLOUD_LOGGING"
    },
    "allocationPolicy": {
        "serviceAccount": {
            "email": "my-sa@my-project.iam.gserviceaccount.com"
        },
        "location": {
            "allowedLocations": [
                "regions/europe-west2"
            ]
        }
    }
}
```

```bash
gcloud batch jobs submit my-job \
	--location europe-west2 \
	--config batch.json
```

Wait for job to fail.

**Comment by averikitsch** [comment]
Thank you for the thorough bug description. I will work on reproducing the issue. In the meantime can you share the error message you are receiving?

**Comment by averikitsch** [comment]
I can reproduce the error, where the email is found from the credentials but not from the metadata server. The fix in v0.2.2 should solve this problem. Please reopen this issue if you continue to experience this issue.
